File: src/main/java/med/voll/api/controller/MedicoController.java
package med.voll.api.controller;

import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import jakarta.validation.Valid;
import med.voll.api.dtos.DadosAtualizacaoMedico;
import med.voll.api.dtos.DadosCadastroMedico;
import med.voll.api.dtos.DadosDetalhamentoMedico;
import med.voll.api.dtos.DadosListagemMedico;
import med.voll.api.services.MedicoService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springdoc.core.annotations.ParameterObject;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.web.PageableDefault;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.util.UriComponentsBuilder;

import java.util.HashMap;
import java.util.Map;

@RestController
@RequestMapping("/medicos")
@SecurityRequirement(name = "bearer-key")
public class MedicoController {

    private static final Logger log = LoggerFactory.getLogger(MedicoController.class);


    @Autowired
    private MedicoService medicoService;


    @PostMapping
    public ResponseEntity<DadosDetalhamentoMedico> cadastrarMedico(@RequestBody @Valid DadosCadastroMedico dadosCadastroMedico, UriComponentsBuilder uriComponentsBuilder) {
        log.info("Recebida solicitação para cadastrar médico: {}", dadosCadastroMedico.nome());
        DadosDetalhamentoMedico dadosDetalhamentoMedico = medicoService.cadastrar(dadosCadastroMedico);
        var uri = uriComponentsBuilder.path("/medicos/{id}").buildAndExpand(dadosDetalhamentoMedico.id()).toUri();
        return ResponseEntity.created(uri).body(dadosDetalhamentoMedico);
    }


    @GetMapping
    public ResponseEntity<?> listarMedicos(@Parameter(hidden = true) @ParameterObject @PageableDefault(page = 0, size = 10, sort = "nome") Pageable paginacao) {
        log.info("Recebida solicitação para listar médicos com paginação: {}", paginacao);

        Page<DadosListagemMedico> dadosListagemMedicos = medicoService.listar(paginacao);

        if (dadosListagemMedicos.isEmpty()) {
            Map<String, String> response = new HashMap<>();
            response.put("message", "Nenhum médico ativo encontrado");
            return ResponseEntity.status(HttpStatus.OK).body(response);
        }

        return ResponseEntity.ok(dadosListagemMedicos);
    }


    @PutMapping
    public ResponseEntity<DadosDetalhamentoMedico> atualizarMedico(@RequestBody @Valid DadosAtualizacaoMedico dadosAtualizacaoMedico) {
        log.info("Recebida solicitação para atualizar médico com ID: {}", dadosAtualizacaoMedico.id());
        DadosDetalhamentoMedico dadosDetalhamentoMedico = medicoService.atualizar(dadosAtualizacaoMedico);
        return ResponseEntity.ok(dadosDetalhamentoMedico);
    }


    @DeleteMapping("/{id}")
    public ResponseEntity<Void> excluirMedico(@PathVariable Long id) {
        log.info("Recebida solicitação para excluir médico com ID: {}", id);
        medicoService.excluir(id);
        return ResponseEntity.noContent().build();
    }


    @GetMapping("/{id}")
    public ResponseEntity<DadosDetalhamentoMedico> detalharMedico(@PathVariable Long id) {
        log.info("Recebida solicitação para detalhar médico com ID: {}", id);
        DadosDetalhamentoMedico dadosDetalhamentoMedico = medicoService.detalhar(id);
        return ResponseEntity.ok(dadosDetalhamentoMedico);
    }
}

File: src/main/kotlin/com/med/scheduler/infrastructure/api/controller/MedicoController.kt
package com.med.scheduler.infrastructure.api.controller

import com.med.scheduler.application.dto.DadosAtualizacaoMedico
import com.med.scheduler.application.dto.DadosCadastroMedico
import com.med.scheduler.application.dto.DadosDetalhamentoMedico
import com.med.scheduler.application.dto.DadosListagemMedico
import com.med.scheduler.application.service.MedicoUseCase
import io.swagger.v3.oas.annotations.Parameter
import io.swagger.v3.oas.annotations.security.SecurityRequirement
import jakarta.validation.Valid
import org.slf4j.LoggerFactory
import org.springdoc.core.annotations.ParameterObject
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.web.PageableDefault
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.util.UriComponentsBuilder

@RestController
@RequestMapping("/medicos")
@SecurityRequirement(name = "bearer-key")
class MedicoController(
    private val medicoUseCase: MedicoUseCase
) {
    private val log = LoggerFactory.getLogger(MedicoController::class.java)

    @PostMapping
    fun cadastrar(
        @RequestBody @Valid dados: DadosCadastroMedico,
        uriBuilder: UriComponentsBuilder
    ): ResponseEntity<DadosDetalhamentoMedico> {
        log.info("Recebida solicitação para cadastrar médico: {}", dados.nome)
        val detalhamento = medicoUseCase.cadastrar(dados)
        val uri = uriBuilder.path("/medicos/{id}").buildAndExpand(detalhamento.id).toUri()
        return ResponseEntity.created(uri).body(detalhamento)
    }

    @GetMapping
    fun listar(
        @Parameter(hidden = true) @ParameterObject
        @PageableDefault(page = 0, size = 10, sort = ["nome"]) paginacao: Pageable
    ): ResponseEntity<Any> {
        log.info("Recebida solicitação para listar médicos com paginação: {}", paginacao)
        val medicos = medicoUseCase.listar(paginacao)

        return if (medicos.isEmpty) {
            ResponseEntity.ok<Any>(mapOf("message" to "Nenhum médico ativo encontrado"))
        } else {
            ResponseEntity.ok<Any>(medicos)
        }
    }

    @PutMapping
    fun atualizar(@RequestBody @Valid dados: DadosAtualizacaoMedico): ResponseEntity<DadosDetalhamentoMedico> {
        log.info("Recebida solicitação para atualizar médico com ID: {}", dados.id)
        return ResponseEntity.ok(medicoUseCase.atualizar(dados))
    }

    @DeleteMapping("/{id}")
    fun excluir(@PathVariable id: Long): ResponseEntity<Void> {
        log.info("Recebida solicitação para excluir médico com ID: {}", id)
        medicoUseCase.excluir(id)
        return ResponseEntity.noContent().build()
    }

    @GetMapping("/{id}")
    fun detalhar(@PathVariable id: Long): ResponseEntity<DadosDetalhamentoMedico> {
        log.info("Recebida solicitação para detalhar médico com ID: {}", id)
        return ResponseEntity.ok(medicoUseCase.detalhar(id))
    }
}
